tools/wake/config.cpp

```diff
@@ -0,0 +1,256 @@
+    // seems like an ok trade off.
+    std::string shell_string = "echo -n " + to_expand;
+
+    // overwrite stdout with the write side of the pipe
+    if (dup2(stdoutPipe[write_side], STDOUT_FILENO) == -1) {
+      perror("Failed to dup2 stdout pipe for eviction process");
+      exit(EXIT_FAILURE);
+    }
+
+    // We don't want stdin/stderr messing things up and
+    // we no longer need the write/read side of the pipe
+    // since stdout *is* the write side of the pipe now.
+    close(STDIN_FILENO);
+    // close(STDERR_FILENO);
+    close(stdoutPipe[read_side]);
+    close(stdoutPipe[write_side]);
+
+    // Now replace this process with the shell command but
+    // exit and fail if we don't manage to do that.
+    execl("/bin/sh", "/bin/sh", "-c", shell_string.c_str(), nullptr);
+    perror("Failed to exec /bin/sh");
+    exit(EXIT_FAILURE);
+  }
+
+  // Make sure to close this side of the pipe so that the shell program doesn't
+  // stay open waiting for commands forever.
+  close(stdoutPipe[write_side]);
+
+  // Read everything the child process outputs and append
+  // it to a string as we go.
+  std::string acc;
+  char buf[4096];
+  while (true) {
+    ssize_t size = read(stdoutPipe[read_side], buf, sizeof(buf));
+    if (size < 0 && errno == EINTR) {
+      continue;
+    }
+    if (size < 0) {
+      perror("failed to exec /bin/sh");
+      exit(EXIT_FAILURE);
+    }
+    if (size == 0) {
+      break;
+    }
+    acc.append(buf, buf + size);
+  }
+
+  // We should be done already but let's wait for the child and
+  // make sure everything worked well there.
+  int status = 1;
+  if (waitpid(pid, &status, 0) != pid) {
+    perror("waitpid: failed to wait /bin/sh -c echo ...");
+    exit(EXIT_FAILURE);
+  }
+  if (status != 0) {
+    fprintf(stderr, "waitpid: /bin/sh -c echo ... failed with non-zero exit status");
+    exit(EXIT_FAILURE);
+  }
+
+  // Return the string we got back
+  return acc;
+}
+
+// Find the default location for the user level wake config
+std::string default_user_config() {
+  // If XDG_CONFIG_HOME is set use it, otherwise use home
+  std::string prefix = shell_expand("$XDG_CONFIG_HOME");
+  if (prefix == "") {
+    prefix = "~";
+  }
+  return prefix + "/.wake.json";
+}
+
+wcl::optional<JAST> read_json_file(const std::string& path) {
+  std::ifstream file(path);
+  if (!file) {
+    // TODO: add this error message back once I have a result type
+    // std::cerr << "Failed to read '" << path << "'" << std::endl;
+    return {};
+  }
+
+  std::stringstream buff;
+  buff << file.rdbuf();
+  std::string contents = buff.str();
+
+  JAST json;
+  std::stringstream errors;
+  if (!JAST::parse(contents, errors, json)) {
+    std::cerr << path << " must be a valid JSON object: " << errors.str() << std::endl;
+    return {};
+  }
+
+  return {wcl::in_place_t{}, std::move(json)};
+}
+
+bool insert_top_level_values(std::unordered_map<std::string, std::string>& map,
```

Comment:
I think we should parse into a known struct so that each field has its own parsing, not a catch all. It should be valid for some configs to be objects. For instance you want to shell expand the user_config but not the version already.